### PR TITLE
Map and SetBuilders [ci: last-only]

### PR DESCRIPTION
**Pull request by mkeskells**
Initial Map/SetBuilder
mostly a placeholder for additional optimizations on map and set building
without breaking binary compatibility

initial optimisation for empty + collection and size

### CODE REVIEW

File: src/library/scala/collection/immutable/MapBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  def result(): C = elems
+
+  final def size: Int = elems.size
+  final def isEmpty: Boolean = elems.isEmpty
+
+  /**
+    * are the types implementationally compatible, i.e value isInstanceOf[C]
+    */
+  protected def isCompatibleType(value: IterableOnce[(K, V)]) : Boolean
```
Comment (julienrf) [inline src/library/scala/collection/immutable/MapBuilder.scala]: What about having a method returning an `Option[C]` instead?

**Review by julienrf — COMMENTED** [review]
What kind of optimizations do you have in mind? Maybe something that would make computing the union of two sets/maps faster?

File: src/library/scala/collection/immutable/MapBuilder.scala (hunk 1 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/MapBuilder.scala]: @julienrf conceptually, but trying to reduce allocations. Maybe its a partial function


Could do with a better name though

**Comment by mkeskells** [comment]
@julienrf I had a PR a while ago to optimise the merge of maps, but stopped work on it due to the new collections

There are lots of simple optimisations that can be applied to addAll in a similar way to merge for a HashMap

Additionally maintaining a Map in immutable form creates object churn that is not needed (that's why this PR makes the Builder break from ImmutableBuilder)

File: src/library/scala/collection/immutable/MapBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  override def addAll(xs: IterableOnce[(K, V)]): this.type = {
+    if ((elems eq empty) && isCompatibleType(xs)) {
```
Comment (szeiger) [inline src/library/scala/collection/immutable/MapBuilder.scala]: Nothing in `MapBuilder`, in particular this optimization, looks Map-specific. The same goes for `SetBuilder`. Would it make more sense to include these changes in `ImmutableBuilder`?

**Comment by mkeskells** [comment]
@szeiger I wanted to enable optimisations that are map and set specific. That would break binary compatibility after 2.13.0 is cut, e.g. merging a Map, so the purpose of this PR is to have an API that enables these additional methods to be available, so that we can extends to implementation later without changing the signature

I am happy to make the corresponding changes in ImmutableBuilder, but also wanted to discuss what APIs should be in builder. Didn't want to make too many changes in different areas and lose the focus

**Comment by julienrf** [comment]
I guess we can always add specialized builder classes after 2.13 is released as long as these classes are private? Or do you want to be able to use it from outside of the standard library?

**Comment by mkeskells** [comment]
@julienrf there is no reason why a builder for a Map is a wrapper around an immutable Map - (that why it should be a different hierarchy) ImmutableBuilder is a [current, and inefficient] implementation detail. 

I am not intending to do anything outside the standard library, just to improve the performance of the library

**Comment by julienrf** [comment]
@mkeskells Yeah, I agree that there must be more efficient builder structures. That being said, I don’t think the current design prevents us to add more efficient `Map` and `Set` builders afterwards in binary compatible way.

**Comment by szeiger** [comment]
As long as the `newBuilder` methods return a `Builder` and not something more specific like `ImmutableBuilder` or `MapBuilder` it should be possible to change the implementation later by adding a new private builder class without breaking binary compatibility.

**Comment by mkeskells** [comment]
@szeiger this overlaps with our previous discussions. 

There are a multitude of additional possible methods that a builder could expose
++, --, size, intersect, union ....  - what should be exposed is a separate concern

allowing a specialised return type enables these methods to be exposed, albeit with binary compatability staged after the initial build is cut

**Comment by szeiger** [comment]
A more specific return type is not sufficient. We cannot add any non-private methods in minor releases due to forward compatibility.

We can override an existing method in an existing or new implementation class but this actually requires that the implementation class is *not* exposed to callers through the static type, otherwise calls of the new override will be made against the implementation class's symbol (not the original definition) which also breaks forward binary compatibility. This was the mistake we made in 2.12.5.

**Comment by szeiger** [comment]
On second thought, I think that's not quite correct. Adding an override in an existing implementation class should be OK. Code compiled against the new version will use the implementation class's symbol but the JVM can resolve that correctly as long as the class already existed in the old version, even if it didn't have the override.

**Comment by szeiger** [comment]
I just took another look at this. Here's what I think we should do:

- Put the code that is currently duplicated in `SetBuilder` and `MapBuilder` here directly into `ImmutableBuilder`. The class is new in 2.13 so there are no compatibility constraints. In fact, some parts are already there. What we need is `size`, `isEmpty` and the `addAll` optimization.

- Create the subclasses `MapBuilder` and `SetBuilder` for the standard map and set kinds. They should implement `addOne` and also override `addAll` to fall back to `concat` for more efficient appending. Both only depend on the kind but not the exact type of the collection. If you add a `ClassTag` for `C` you can also implement the type check in a generic way so there's no need for map and set implementations to subclass these builders. Non-standard-kinded collections (`LongMap`, `TreeMap`, `TreeSet`) need their own specific `ImmutableBuilder` subclasses.

- But why should only builders benefit from the concat-to-empty optimization? We could bake this directly into the `empty` objects by overriding their `concat` methods. This would also simplify `ImmutableBuilder` (i.e. no `ClassTag` necessary, `addAll` can always call `concat`).

**Comment by mkeskells** [comment]
I should have some time late weekend and next week to catch up on this backlog.
I will progress these ideas ASAP

**Comment by mkeskells** [comment]
@szeiger we cant add size and isEmpty to ImmutableBuilder without changing the declaration of ImmutableBuilder - `elems` has protected access, and `size` is not available on `C`, so I cant do size and isEmpty

I have made someother changes in line ( I think)  with your suggestions 
Not sure what you expected to  wit the empty concat. Whas that for this PR?

**Comment by SethTisue** [comment]
needs rebase (because conflicts, but also, doing so should un-confuse Scabot)

**Comment by mkeskells** [comment]
@szeiger I think we will need to tighter specif the empty objects, so we can delegate
code like this fails
```
 @Test
  def simpleBitset(): Unit = {
    val b = Set.newBuilder[Any]
    b ++= BitSet(1)
    val set:Set[Any] = b.result()
    val more = set + "fish"
  }
```
`elems` of the builder is narrowed by the ++ to be a Bitset, then the add fails
```
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
	at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:101)
	at scala.collection.immutable.BitSet.incl(BitSet.scala:19)
	at scala.collection.immutable.SetOps.$plus(Set.scala:35)
	at scala.collection.immutable.SetOps.$plus$(Set.scala:35)
	at scala.collection.immutable.AbstractSet.$plus(Set.scala:250)
	at scala.collection.immutable.SetTest.simpleBitset(SetTest.scala:87)
```

File: src/library/scala/collection/immutable/MapBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+    * are the types implementationally compatible
+    */
+  protected def isCompatibleType(value: IterableOnce[(K, V)]): Boolean =
+    classTag[C].runtimeClass.isInstance(value)
+
+  def addOne(elem: (K, V)): this.type = {
+    elems = (elems + elem).asInstanceOf[C]; this
+  }
```
Comment (retronym) [inline src/library/scala/collection/immutable/MapBuilder.scala]: Should we have `def addKeyValue(k: K, v: V)` here? to allow construction without tupling?

File: src/library/scala/collection/immutable/MapBuilder.scala [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  override def addAll(xs: IterableOnce[(K, V)]): this.type = {
+    if ((elems eq empty) && isCompatibleType(xs)) {
+      elems = xs.asInstanceOf[C]
```
Comment (retronym) [inline src/library/scala/collection/immutable/MapBuilder.scala]: I don't think this is safe for a few reasons.

First, it aliases the incoming, potentially mutable collection, which might be later mutated by the caller.

Second, if `xs` was a `MyFancyMap` that  extended `MyMap`, I'd be surprised not to exactly get back a `MyMap` from this builder.

File: src/library/scala/collection/immutable/MapBuilder.scala (hunk 3 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/MapBuilder.scala]: done

File: src/library/scala/collection/immutable/MapBuilder.scala (hunk 4 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/MapBuilder.scala]: The first case is covered I think - the incoming could not be an immutable.Set/Map

The extension is similar to the issue that I posted above with BitSet I think

**Comment by joshlemer** [comment]
Could we get away with having a quite a bit less restrictive `MapBuilder` and `SetBuilder`? Something like

```scala
trait SizedBuilder[-A, C] {
  def size: Int
  def isEmpty: Boolean
}

trait MapBuilder[K, V, M[K0, V0] <: Map[K0, V0]]
  extends mutable.ReusableBuilder[(K, V), M[K, V]] {
  def addKeyValue(key: K, value: V): this.type
}

trait SetBuilder[A, S[A0] <: Set[A0]] 
  extends mutable.ReusableBuilder[A, S[A0]] {
  // ...
}
```

should provide all the required methods to create more optimized implementations, without forcing that on all impl's. For instance, `MapBuilder`'s `protected var elems: C = empty` restricts all implementations to manage their elements as an `immutable.map[K,V]`, but surely some implementations could want to do something else.

**Comment by mkeskells** [comment]
@joshlemer seems sensible, but would require some input from core commiters abut what abstractions they want to expose. My expectation is that we can get improvements from having mutability in the content during the build phase
@szeiger @retronym what do you think

**Pull request closed without merging**